django/db/models/base.py

```diff
@@ -1796,6 +1798,22 @@ def _check_field_name_clashes(cls):
             used_fields[f.name] = f
             used_fields[f.attname] = f
 
+        # Additional check for OneToOnefield
+        # Not listed in cls._meta.local_fields
+        for local_f in cls._meta.parents.values():
+            if not local_f:
+                continue
+            clash = parent_fields.get(local_f.name) or None
+            if clash:
+                errors.append(
+                    checks.Error(
+                        "The field '%s' clashes with the field '%s' "
+                        "from model '%s'."
```

Comment:
Use f-string.

Reply:
We could avoid creating `parent_fields` by:
```diff
diff --git a/django/db/models/base.py b/django/db/models/base.py
index 8047a11afd..ca00149500 100644
--- a/django/db/models/base.py
+++ b/django/db/models/base.py
@@ -1770,13 +1770,26 @@ class Model(AltersData, metaclass=ModelBase):
         # Check that fields defined in the model don't clash with fields from
         # parents, including auto-generated fields like multi-table inheritance
         # child accessors.
-        parent_fields = {}
         for parent in cls._meta.get_parent_list():
             for f in parent._meta.get_fields():
-                parent_fields[f.name] = f
                 if f not in used_fields:
                     used_fields[f.name] = f
 
+        for parent_link in cls._meta.parents.values():
+            if not parent_link:
+                continue
+            clash = used_fields.get(parent_link.name) or None
+            if clash:
+                errors.append(
+                    checks.Error(
+                        f"The field '{parent_link.name}' clashes with the field '{clash.name}' "
+                        f"from model '{clash.model._meta}'.",
+                        obj=parent_link,
+                        id="models.E006",
+                    )
+                )
+
         for f in cls._meta.local_fields:
             clash = used_fields.get(f.name) or used_fields.get(f.attname) or None
             # Note that we may detect clash between user-defined non-unique
@@ -1798,22 +1811,6 @@ class Model(AltersData, metaclass=ModelBase):
             used_fields[f.name] = f
             used_fields[f.attname] = f
 
-        # Additional check for OneToOnefield
-        # Not listed in cls._meta.local_fields
-        for local_f in cls._meta.parents.values():
-            if not local_f:
-                continue
-            clash = parent_fields.get(local_f.name) or None
-            if clash:
-                errors.append(
-                    checks.Error(
-                        "The field '%s' clashes with the field '%s' "
-                        "from model '%s'."
-                        % (local_f.name, clash.name, clash.model._meta),
-                        obj=local_f,
-                        id="models.E006",
-                    )
-                )
         return errors
```

Reply:
Please check it again, it doesn't cause any failures for me. It must be moved before checking local fields as it mutates `used_fields`.

Reply:
A simple hack would be to add another check if the `hasattr(parent_link, attname)` is True like the following, but that won't be standard practice I think
```
if not parent_link or hasattr(parent_link, 'attname'):
      continue
```

Reply:
I totally missed to check this! Thank you. I was putting it at the end blindly from the beginning and at last decided to create a new dict.

Reply:
Yes, so just to be clear, this is what I'm proposing to apply to the current `main` branch 
```diff
diff --git a/django/db/models/base.py b/django/db/models/base.py
index 959b72c93b..1aa6ff1d2d 100644
--- a/django/db/models/base.py
+++ b/django/db/models/base.py
@@ -1775,6 +1775,20 @@ class Model(AltersData, metaclass=ModelBase):
                 if f not in used_fields:
                     used_fields[f.name] = f
 
+        for parent_link in cls._meta.parents.values():
+            if not parent_link:
+                continue
+            clash = used_fields.get(parent_link.name) or None
+            if clash:
+                errors.append(
+                    checks.Error(
+                        f"The field '{parent_link.name}' clashes with the field "
+                        f"'{clash.name}' from model '{clash.model._meta}'.",
+                        obj=parent_link,
+                        id="models.E006",
+                    )
+                )
+
         for f in cls._meta.local_fields:
             clash = used_fields.get(f.name) or used_fields.get(f.attname) or None
             # Note that we may detect clash between user-defined non-unique

```

Reply:
Just a little skeptical about the `obj=parent_link` part as the error message does not identifies the conflicted model and looks like the following  https://github.com/django/django/pull/16952#discussion_r1233827427